README.md




<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://www.localdataexchange.com">
    <img src="https://staging-ipromote.ldex.co/ctm/LDE_Logo-Black.png" alt="Logo" width="" height="80">
  </a>

  <h3 align="center">Api Helper Package</h3>

  <p align="center">
    A package to consume api smoothly
    <br />
    <a href="#table-of-contents"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://packagist.org/packages/lde/api-helper">View Package</a>
    ·
    <a href="https://github.com/Local-Data-Exchange/api-helper/issues">Report Bug</a>
    ·
    <a href="https://github.com/Local-Data-Exchange/api-helper/issues">Request Feature</a>
  </p>
</p>



<!-- TABLE OF CONTENTS -->
## Table of Contents

* [Getting Started](#getting-started)
* [Installation](#installation)
* [Configuration](#configuration)
* [Usage](#usage)
	* [Methods](#methods)
* [Response](#response)
		



## Getting Started    

This package is useful to consume API's, here is the instruction for installation and usage.

## Installation
   
1. To install this package using [Packagist](https://packagist.org/packages/lde/api-helper)  

2. On the root of your project run following command   

		composer require lde/api-helper

3. This command will install package with dependency
  
## Configuration

- To use this apihelper need to export config file to do so run the following command in your terminal to publish config file to config folder.

	    php artisan vendor:publish  --provider="Lde\ApiHelper\ApiHelperServiceProvider"

- This will publish config file naming **api_helper.php** into config folder.

## Prometheus Configuration
- Prometheus is dependent on your app so you need to provide prometheus configuration and also use below packages on your app.

  - [jimdo/prometheus_client_php](https://github.com/Jimdo/prometheus_client_php)
  - [superbalist/laravel-prometheus-exporter](https://github.com/Superbalist/laravel-prometheus-exporter)
- If you want to use prometheus then you should turn it on from config [api_helper](src/Config/api_helper.php)
```php
'log_stats' => true, // If you want to use prometheus then set as true otherwise false

    'prometheus' => [
        'labels' => [           
            'client_id' => 10,
            'app' => 'api-helper',
            'source' => 'core',
        ],
        'histogram_bucket' => [0.1, 0.25, 0.5, 0.75, 1.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 7.5, 10.0],
    ],
```
- You can configure labels of prometheus inside `prometheus.labels` as per your need.
- `histogram_bucket` you can set inside prometheus config as array.

## Usage

- To use this package you need to add following class where you want to use this package.

		use Lde\ApiHelper\ApiBuilder;

		
### Methods

#### addHeaders($headers)

- This method is use to add headers.

- It accept name and value as parameter, Here you can set only one header at a time.


		$headers['Accept'] = "application/json"; 
		$headers['Content-Type'] = "application/json";  
		app(ApiBuilder::class)->addHeaders($headers);

- We will get response in form of object of ApiBuilder.


#### api($connection)

- This method is use to set api that we are going to use from *api_helper.php* , there is httpbin and mokbin is define so you have to pass the name that you want to use.

- You can also define your own api end point at *api_helper.php* in config file.
	
        app(ApiBuilder::class)->api('httpbin')->method_to_call();

- The snippet indicates how you can connect particular api and access their method.

- method_to_call() is the function that you have specified inside *api_helper* connection array.

- This will return object of ApiResponse.

### Response

- Here you will get object in response, In each response you will get success either true or false
- You will also get status code for more information about response please check below doc.
- http://docs.guzzlephp.org/en/latest/psr7.html#responses


CHANGELOG.md
# CHANGELOG

## 7.0.0

**Upgrade package to support laravel 7** [Ref](https://laravel.com/docs/7.x/upgrade)

- Upgrade packages to support laravel 7.
- php `^7.2.5` or higher
- array-to-xml `~2.14.0`,
- php-dot-notation `^2.x.dev`
- phpunit `8.5` 


## 6.0.0

**Upgrade package to support laravel 6**

- Syantaxt upgraded to support laravel 6.*


## 0.1.30
- Fixed return type issue when json_decode fails. [Hotfix]

## 0.1.29
- Fixed json mapping condition for empty body. [Hotfix]

## 0.1.28
- Perform json mapping if body is given.

## 0.1.27
- Bug fix of xml boolean assign.

## 0.1.26
- Adding support of boolean in query param.

## 0.1.25
- Add support of optional query param.

## 0.1.24
- Adding logs in json mapping, so we can figure out if something is wrong with data.

## 0.1.23 
- Added form_params support.
- Update circle ci config to test with laravel 6.
- Update circle ci to run testcases smoothly and fast.

## 0.1.22
- This version include support of call back function for prometheus config.

## 0.1.21
- Add support for metrics, which will use app prometheus to export metrics.

## 0.1.20
- Add option to add status code to not retry, this option you can set inside your api config.

## 0.1.19
- Bump symfony/http-foundation from 4.2.6 to 4.4.1
- Remove space from `&amp;` inside processXmlMappings fucntion.

## 0.1.18
- Updated document and added new docker image files in config for testing different environment.
- When value is empty or null at that time we were sending UNKNOWN which is now replaced with null in process json mappings function.
- For process xml mappings, when it have empty or null values we are removing from that string. Instead of returning UNKNOWN.

## 0.1.17
- Added new ability in json mappings, now it support nullable fields, which can be define in api_helper.

## 0.1.16
- Added variable type checking in checkbool method, it will only check if it is string.

## 0.1.15
- Callable method return false due to invalid method name.

## 0.1.14
- Bug solved of variable naming convention.

## 0.1.13
- Added custom method support for escaping string in processXmlMappings methods.
- Added parameter in api_helper config file which defines custom escape string method of user.

## 0.1.12
- Remove space from processXmlMappings methods.

## 0.1.11
- Converted output of simplexml_load_string into array.

## 0.1.10
- Remove constructor, also removed redundant code for improvements.

## 0.1.9
- Bug Fixed guzzle http response show 200 even it is fail.

## 0.1.8
- Added new logic for combining additional headers, and deafult headers.
 
## 0.1.7
- Due to default headers there is possibility of duplicate headers that is fixed in this version.

## 0.1.6
- Applied solution in version 0.1.5 not working properly so now, it is updated with different code flow.

## 0.1.5
- Fixing header has been overwritten by api method, when headers is being set using addHeaders.

## 0.1.4
- Added validation in escapeSpecialCharacters() and checkBool() to return value if it is not empty.

## 0.1.3
- Upgrade laravel/framework version, because incompatible with orchestra/testbench package.
- Downgrade compatible version of adbario/php-dot-notation from "2.x-dev" to "^2.2.0".
- Added unit testing with circle ci.

## 0.1.2
- Updated README.md file with new document structure.
- Remove addHeader() from ApiBuilder.
- Added new dependency in composer.json file for the package.
- Update api_helper.php config file for root elements.
- Updated call() and xmlParsing() to get root element from the config file.
- Added phpunit test cases in "tests" folder.
- Created phpunit.xml file so test cases can be run on any cloud server.
- Added orchestra/testbench with phpunit for package testing.
- Updated TestCase added methods for provider and alias, also now able to mock config.

## 0.1.1
- Added ApiResponse.php now response will be object of ApiResponse.
- Auto package discovery added, now no need to register for provider those who have laravel version >= 5.5 and having dev environment.

## 0.1.0
- Helpers, Config, Events added for ApiBuilder.php .
- Created Facade and Service provide for package.



src/Config/api_helper.php
<?php

return [
    /*
    |--------------------------------------------------------------------------
    | Prometheus
    |--------------------------------------------------------------------------
    | log_stats if you want to use api stats using prometheus
    |
    | log_stats = false (default)
    |
    | Set config of prometheus like app name, client etc.
    |
    |
    | You can any variables to export on histogram and counter of prometheus
     */
    'log_stats' => true,

    'prometheus' => [
        'labels' => [
            'client_id' => 10,
            'app' => 'api-helper',
            'source' => 'core',
        ],
        'histogram_bucket_name' => 'lde_http_response_time_seconds_bucket',
        'histogram_bucket' => [0.1, 0.25, 0.5, 0.75, 1.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 7.5, 10.0],
    ],

    /*
    |--------------------------------------------------------------------------
    | Retries
    |--------------------------------------------------------------------------
    |
    | Sets the number of retries to attempt before giving up.
    |
    | 0 = Off
    |
     */

    'number_of_retries' => 3,

    /*
    |--------------------------------------------------------------------------
    | Default Guzzle Request Options
    |--------------------------------------------------------------------------
    |
    | Sets the default Guzzle request options.
    |
    | http://docs.guzzlephp.org/en/stable/request-options.html
    |
    | 0 = Off
    |
     */

    'default_request_options' => [
        'http_errors' => true,
        'connect_timeout' => 10,
        'timeout' => 30,
        'headers' => [
            "Accept" => "application/json",
            "Content-Type" => "application/json",
        ],
    ],

    // If no connection provided, use default
    'default' => 'httpbin',

    // Sensitive fields will be masked on the return value and logs
    'sensitive_fields' => [
        'auth.0',
        'auth.1',
        'headers.apikey',
    ],

    // Define all the APIs here
    'connections' => [

        // HTTPBin
        'httpbin' => [
            'root' => '',

            //Set method name for custom escape characters
            'character_escape_method' => '',

            // API type: json or xml or view
            'type' => 'json',

            // API base URL
            'base_url' => 'https://httpbin.org',

            // Default Request options. these are included with all calls unless overwritten
            'default_request_options' => [
                'http_errors' => true,
                'connect_timeout' => 10,
                'timeout' => 30,
                'headers' => [
                    "Accept" => "application/json",
                    "Content-Type" => "application/json",
                ],
            ],

            // number of retries to override global settings.
            'number_of_retries' => 0, // 0 = off

            //status not to retry, define status code for which api will not retried
            'status_not_to_retry' => [517],

            // List of API routes we are integrating with
            'routes' => [

                // Sample API to test GET
                'get' => [
                    'method' => 'GET',
                    'uri' => '/get',
                    'mappings' => [
                        'path' => [],
                        'query' => [
                            'name' => 'person.name',
                            'surname' => 'person.surname',
                            'foo' => 'foo',
                        ],
                        'body' => [],
                    ],
                ],

                // Sample API to test POST
                'post' => [
                    'name' => 'httpbin',
                    'method' => 'POST',
                    'uri' => '/post',
                    'body' => [
                        'first_name' => '{name}',
                        'last_name' => '{surname}',
                        'nested' => [
                            'foo' => '{foo}',
                        ],
                    ],
                    'mappings' => [
                        'query' => [
                            'test' => 'person.name',
                        ],
                        'body' => [
                            'name' => 'person.name',
                            'surname' => 'person.surname',
                            'foo' => 'foo',
                        ],
                    ],
                ],
                // Sample API to test Form_params

                'formParams' => [
                    'name' => 'httpbin',
                    'method' => 'POST',
                    'uri' => '/post',
                    'request_type' => 'form_data',
                    'form_params' => [
                        'parameters' => [
                            'revision' => '{revision}',
                        ],
                    ],
                    'mappings' => [
                        'form_params' => [
                            'revision' => 'param.revision',
                        ],
                    ],
                ],
                'delete' => [
                    'name' => 'httpbin',
                    'method' => 'DELETE',
                    'uri' => '/delete',
                    'mappings' => [
                        'query' => [
                            'id' => 'person.id',
                        ],
                    ],
                ],

            ],
        ],

        'mockbin' => [
            'root' => 'request',

            //Set method name for custom escape characters
            'character_escape_method' => '',

            // API type: json or xml or view
            'type' => 'xml',

            // API base URL
            'base_url' => 'http://mockbin.org',

            // Default Request options. these are included with all calls unless overwritten
            'default_request_options' => [
                'http_errors' => true,
                'connect_timeout' => 10,
                'timeout' => 30,
                'headers' => [
                    "Accept" => "application/xml",
                    "Content-Type" => "application/xml",
                ],
            ],

            // number of retries to override global settings.
            'number_of_retries' => 0, // 0 = off

            'routes' => [
                // Sample API to test POST using xml as the body
                'echo' => [
                    'method' => 'POST',
                    'uri' => '/echo',
                    'request_options' => [ // we can override request_options per API
                        'http_errors' => true,
                        'connect_timeout' => 10,
                        'timeout' => 30,
                        'headers' => [ // these headers will be set automatically for XML apis. No need to specify them
                            "Accept" => "application/xml",
                            "Content-Type" => "application/xml",
                        ],
                    ],
                    'xml_config' => [
                        'root_element_name' => 'request', // defaults to root if left out
                        'attributes' => [
                            'xmlns' => 'https://github.com/spatie/array-to-xml',
                        ],
                        'use_underscores' => true,
                        'encoding' => 'UTF8',
                    ],
                    'body' => [
                        'request' => [
                            'attributes' => ['class' => '{class}'],
                            'name' => '{name}',
                            'weapon' => '{weapon}',
                        ],
                    ],
                    'mappings' => [
                        'body' => [
                            'class' => 'request.class',
                            'name' => 'request.name',
                            'weapon' => 'request.weapon',
                        ],
                    ],
                ],
            ],
        ],
    ],
];
